ha/core/event_manager/event_manager.py

```diff
@@ -37,62 +49,263 @@ def get_instance():
+
+    @staticmethod
+    def _validate_component(component: str = None) -> None:
         """
         Validate component raise error invalid component.
 
         Args:
             component (str): Component name
         """
-        if component.lower() not in SUBSCRIPTION_LIST:
+        if component is None or component.lower() not in SUBSCRIPTION_LIST:
             raise InvalidComponent(f"Invalid component {component}, not part of subscription list.")
 
-    def _validate_events(self, events: list) -> None:
-        """
-        Raise error InvalidEvent if event is not valid.
+    def subscribe(self, component: str = None, events: list = None):
+        '''
+           Register all the events for the notification. It maintains
+           list of events registered by the components. It maps the event
+           name with component name and store it in the consul. \
+           Also, maps the component with event
 
-        Args:
-            events (list): Event list.
-        """
-        if not isinstance(events, list):
-            raise InvalidEvent(f"Invalid type {events}, event type should be list")
+           Args:
+            component (str): Component Name
+            events (list): Events.
+        '''
+        # TODO: Add health monitor rules
+        Log.info(f'Received a subscribe request from {component}')
+        EventManager._validate_component(component)
+        EventManager._validate_events(events)
+        if component:
+            message_type = self._create_message_type(component)
         for event in events:
-            if event not in EVENTS:
-                raise InvalidEvent(f"Invalid event: {event}, not part of HA event list.")
-        Log.debug(f"event: {event} is valid for subscription request")
+            self._store_component_key(f'{const.COMPONENT_KEY}/{component}', event.resource_type, event.states)
+            self._store_event_key(f'{const.EVENT_KEY}', event.resource_type, event.states, comp=component)
+        Log.info('Subscription request is successfully stored into confstore')
+        return message_type
 
-    def subscribe(self, component: str, events: list) -> str:
-        """
-        Register events for the notification. It maintains list of events registered by the components.
+    def _store_component_key(self, key: str, resource_type: str, val: list) -> None:
+        '''
+           Perform actual consul store operation for component keys
+           Ex:
+           key: /cortx/ha/v1/events/subscribe/sspl
+           confstore value: ['enclosure:sensor:voltage/failed', ...]
+        '''
+        if self._confstore.key_exists(key):
+            comp_json_list = self._confstore.get(key)
 
-        Args:
-            component (str): Component Name
-            events (list): Events.
-        """
-        pass
+            # Get the compnent list from consul first
+            _, comp_list = comp_json_list.popitem()
```

Comment:
I guess it is a list of events, so lets rename to "event_list"

Reply:
Done